services/statestorage/adapter/persistence.go

```diff
@@ -6,6 +6,6 @@ import (
 )
 
 type StatePersistence interface {
-	WriteState(contract primitives.ContractName, stateDiffs *protocol.StateRecord) error // TODO: change this to an array as well since we do multiple writes in one transactions
-	ReadState(contract primitives.ContractName) (map[string]*protocol.StateRecord, error)
+	WriteState(height primitives.BlockHeight, contract primitives.ContractName, stateDiffs *protocol.StateRecord) error // TODO: change this to an array as well since we do multiple writes in one transactions
```

Comment:
Since `WriteState` should always work in the granularity of a block, what happens if we start failing mid-way through a block? The current interface isn't easily transactional

Reply:
It will never be transactional. A better interface will only give us the illusion of transactionality. I do agree, though.

Reply:
interface changed. (internal is still not transactional for now)